src/contrib/mlir/backend/pass/affine_lowerer.cpp

```diff
@@ -537,9 +603,548 @@ namespace
+    }
+
+    void DialectLoweringPass::insertInitFunc()
+    {
+        auto module = getModule();
+        OpBuilder builder(module.getContext());
+        OpBuilder::InsertionGuard moduleInsertionGuard(builder);
+        builder.setInsertionPointToStart(module.getBody());
+
+        // Insert init function
+        auto funcTy = builder.getFunctionType({}, {});
+        SmallVector<NamedAttribute, 4> attributes;
+        auto funcOp = builder.create<mlir::FuncOp>(
+            builder.getUnknownLoc(), "callback_init", funcTy, attributes);
+        // Insert entry block
+        auto entry = funcOp.addEntryBlock();
+        builder.setInsertionPointToStart(entry);
+
+        if (m_attrsVec.size() == 0)
+        {
+            // No callbacks, just return
+            builder.create<mlir::ReturnOp>(builder.getUnknownLoc());
+            return;
+        }
+
+        // Insert operations into entry block
+        auto* llvmDialect = module.getContext()->getRegisteredDialect<mlir::LLVM::LLVMDialect>();
+        auto llvmI64Ty = LLVM::LLVMType::getInt64Ty(llvmDialect);
+        auto llvmI64PtrTy = llvmI64Ty.getPointerTo();
+        auto llvmI32Ty = LLVM::LLVMType::getInt32Ty(llvmDialect);
+        auto llvmI8Ty = LLVM::LLVMType::getInt8Ty(llvmDialect);
+        auto llvmI8PtrTy = llvmI8Ty.getPointerTo();
+        auto llvmArray1DI64Ty = LLVM::LLVMType::getArrayTy(llvmI64Ty, 1);
+        auto llvmArray2DI64Ty = LLVM::LLVMType::getArrayTy(llvmI64Ty, 2);
+        auto llvmArray3DI64Ty = LLVM::LLVMType::getArrayTy(llvmI64Ty, 3);
+        auto llvmF32Ty = LLVM::LLVMType::getFloatTy(llvmDialect);
+
+        // constants needed by gep
+        // LLVM requires that structure indexes be (vectors of) 32-bit integer constants.
+        std::vector<LLVM::ConstantOp> constants;
+        auto maxNumElem = 11;
+        for (auto i = 0; i < maxNumElem; i++)
+        {
+            auto constant = builder.create<LLVM::ConstantOp>(
+                builder.getUnknownLoc(), llvmI32Ty, builder.getI32IntegerAttr(i));
+            constants.push_back(constant);
+        }
+        auto globalType = getLLVMType(AttrsType::CONV3D, llvmDialect);
+        auto gepTy = getLLVMType(AttrsType::CONV3D, llvmDialect).getPointerTo();
+
+        int32_t i = 0;
+        for (auto attrs : m_attrsVec)
+        {
+            StringRef name = "globalAttrs" + std::to_string(i);
+            LLVM::GlobalOp globalVal = getGlobal(name,
+                                                 globalType,
+                                                 false,
+                                                 LLVM::Linkage::Internal,
+                                                 builder.getZeroAttr(globalType),
+                                                 builder);
+            auto globalPtr = builder.create<LLVM::AddressOfOp>(builder.getUnknownLoc(), globalVal);
+
+            switch (m_attrsTyVec[i])
+            {
+            case AttrsType::INT:
```

Comment:
I think adding an example of how the set of attributes looks like for each op would be very useful. It would help verify if the stores are actually doing the right thing.

Reply:
Do you mean to copy the definitions of those structs from there https://github.com/NervanaSystems/ngraph/blob/66bf59ef9c9291b85862374dd3e99522f33a84f4/src/contrib/mlir/runtime/cpu/callback_utils.hpp#L97 to comment?